modules/standard/FileSystem.chpl

```diff
@@ -109,6 +109,177 @@ proc chown(name: string, uid: int, gid: int) {
+  }
+}
+
+/* Copies the contents and permissions of the file indicated by src into
+   the file or directory dest.  If dest is a directory, will generate an error
+   message.  If metadata is set to true, will also copy the metadata of the
+   file to be copied.  May generate other error messages.
+   src: the source file whose contents and permissions are to be copied
+   dest: the destination file for the contents and permissions.
+   metadata: a boolean indicating whether to copy metadata associated with the
+             source file.
+*/
+proc copy(src: string, dest: string, metadata: bool = false) {
+  var err: syserr = ENOERR;
+  copy(err, src, dest, metadata);
+  if err != ENOERR then ioerror(err, "in copy(" + src + ", " + dest + ")");
+}
+
+/* Copies the contents of the file indicated by src into the file indicated
+   by dest, replacing dest if it already exists (and is different than src).
+   If the dest is not writable or src and dest are the same file, generates
+   an error.  Does not copy metadata.
+   err: a syserr used to indicate if an error occurred
+   src: the source file whose contents are to be copied.
+   dest: the destination of the contents.
+*/
+proc copyFile(out err: syserr, src: string, dest: string) {
+  // This implementation is based off of the python implementation for copyfile,
+  // with some slight differences.  That implementation was found at:
+  // https://bitbucket.org/mirror/cpython/src/c8ce5bca0fcda4307f7ac5d69103ce128a562705/Lib/shutil.py?at=default
+  // I did not look at the other functions in that file, except for copyfileobj
+  // (which copyfile called).
+  if (!exists(src)) {
+    err = ENOENT;
+    // Source didn't exist, we can't copy it.
+    return;
+  }
+  if (isDir(err, src) || isDir(err, dest)) {
+    // If the source is a directory, the user has made a mistake, so return an
+    // error.  The same is true if the destination is a directory.
+    err = EISDIR;
+    return;
+  }
+
+  if (err == ENOENT) {
+    err = ENOERR;
+    // We don't care if dest did not exist before, we'll create or overwrite it
+    // anyways.  We already know src exists.
+  } else if (sameFile(err, src, dest)) {
+    // Check if the files are the same, error if yes
+
+    // Don't need to check if they're the same file when we know dest didn't
+    // exist.
+    err = EINVAL;
+    return;
+    // The second argument is invalid if the two arguments are the same.
+  }
+
+  // Lydia note (03/04/2014): These enclosing curly braces are to avoid the bug
+  // found in test/io/lydia/outArgEarlyExit*.future.  When those futures are
+  // resolved, the braces should be removed.
+  {
+    // Open src for reading, open dest for writing
+    var srcFile = open(src, iomode.r);
+    var destFile = open(dest, iomode.cw);
+    var srcChnl = srcFile.reader();
+    var destChnl = destFile.writer();
+    // read in, write out.
+    var line: string;
+    while (srcChnl.readline(line)) {
+      destChnl.write(line);
+    }
```

Comment:
I'd rather see this copy the file data with a binary read to a fixed-size buffer and then binary writes.
Also, the channels should be opened in binary mode. And you could disable locking on these channels.
Right now, with text mode, if (e.g.) you ran into an invalid unicode character, it would be
replaced by the unicode error character and/or you'd get a fatal error. But that doesn't
make any sense in this context.

If you want it to be faster, we can make it only buffer once (sharing the bytes
read into memory between the two channels). To do that you'd do something like this (in a loop):

srcReader.mark
srcReader.advance( buffer size )
srcReader.beginPeekBuffer to get the current buffer (qio_channel_begin_peek_buffer)
dstWriter.putBuffer with the buffer we got (qio_channel_put_buffer)
srcReader.endPeekBuffer

Some of these routines don't exist with Chapel wrappers now though...